### ISSUE TITLE
Add new algorithms - GATE and TabR

### ISSUE THREAD

**Issue by strukevych** [mljar/mljar-supervised#672]
There are two 2023 models that outperform MLP on all or almost all datasets: 
GATE https://arxiv.org/abs/2207.08548
TabR https://arxiv.org/abs/2307.14338

They also outperformed CatBoost/XGBoost/LightGBM on most datasets.

TabR is also a KNN on steroids, so it could be a good replacement for the NeuralNetwork algorithm in that repository.

We can drop NeuralNetwork or/and KNN from the algorithms after implementation.

On some datasets, they achieved even a 10% boost in accuracy metric on the tuned model compared to other tuned alternatives.

They are suitable for both classification and regression problems, but the best performance boost they achieving on the classification datasets

**Comment by pplonski** [comment]
Amazing results for GATE and TabR. Do they have github repos available with implementation? Have you tried those algorithms?

**Comment by strukevych** [comment]
> Amazing results for GATE and TabR. Do they have github repos available with implementation? Have you tried those algorithms?

Yes, TabR has a single benchmark repo (with implementation included)

As I understood, TabR is done by Yandex (same as CatBoost), and it's a kind of replacement for CatBoost, KNN and MLP at the same time.
They also have a very cool feature named context freezing. It can speed up learning up to x100 for a small cost of accuracy so that it can be used even for such things as an exploration on the data.

GATE has repositories both with benchmarks and several implementations

**Comment by strukevych** [comment]
I will do a prototype with a benchmark on my dataset and will place a PR for these algorithms.

But I will take that only after improving handle_imbalance and feature selection for fair results.

**Comment by pplonski** [comment]
Here is TabR repo https://github.com/yandex-research/tabular-dl-tabr

**Comment by strukevych** [comment]
> Here is TabR repo https://github.com/yandex-research/tabular-dl-tabr

Yes, exactly, it's a benchmarking repo basically :)
